File: server/src/main/java/au/com/codeka/warworlds/server/handlers/Route.java
package au.com.codeka.warworlds.server.handlers;

import com.google.common.base.Preconditions;
import com.google.firebase.database.annotations.Nullable;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.annotation.Nonnull;
import javax.servlet.http.HttpServletRequest;

public class Route {
  private Pattern pattern;
  private Class<? extends RequestHandler> handlerClass;
  private String extraOption;

  public Route(@Nonnull String pattern, Class<? extends RequestHandler> handlerClass) {
    this(pattern, handlerClass, null);
  }

  public Route(
      @Nonnull String pattern,
      Class<? extends RequestHandler> handlerClass,
      String extraOption) {
    this.pattern = Pattern.compile(Preconditions.checkNotNull(pattern));
    this.handlerClass = handlerClass;
    this.extraOption = extraOption;
  }

  /**
   * @return The {@link Matcher} that matches the given request, or null if the request doesn't
   * match.
   */
  @Nullable
  public Matcher matches(HttpServletRequest request) {
    String path = request.getPathInfo();
    if (path == null) {
      path = "/";
    }

    Matcher matcher = pattern.matcher(path);
    if (matcher.matches()) {
      return matcher;
    }
    return null;
  }

  public String getExtraOption() {
    return extraOption;
  }

  public RequestHandler createRequestHandler() throws RequestException {
    try {
      return handlerClass.newInstance();
    } catch (InstantiationException | IllegalAccessException e) {
      throw new RequestException(e);
    }
  }
}

File: server/src/main/kotlin/au/com/codeka/warworlds/server/handlers/Route.kt
package au.com.codeka.warworlds.server.handlers

import java.util.regex.Matcher
import java.util.regex.Pattern
import javax.servlet.http.HttpServletRequest

class Route @JvmOverloads constructor(
    pattern: String,
    private val handlerClass: Class<out RequestHandler>,
    val extraOption: String? = null) {
  private val pattern: Pattern = Pattern.compile(pattern)

  /** @return The [Matcher] that matches the given request, or null if the request doesn't match. */
  fun matches(request: HttpServletRequest): Matcher? {
    var path = request.pathInfo
    if (path == null) {
      path = "/"
    }
    val matcher = pattern.matcher(path)
    return if (matcher.matches()) {
      matcher
    } else null
  }

  fun createRequestHandler(): RequestHandler {
    return handlerClass.newInstance()
  }
}